From 73874b55f1e9aaaa43735e3ae8b98f15b53ba45d Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 11 May 2022 13:04:34 +0300
Subject: [PATCH] [SW-33054] greco : move rr lbw logic to common

As per errata GOYA2_0100, any access to PCIE_MSI region would
result in a freeze. This logic for rr was available in the
legacy flow. Hence minimal logic required to block accesses to
PCIE_MSI region is carried over to common code so that it
is also available in secureboot flow.

Change-Id: I09d06b5d11e89b5eebcca19eaca5a6ebb1f488c3
---

diff --git a/arch/greco/arc/security.c b/arch/greco/arc/security.c
index 8b1fbc3..a24c37e 100644
--- a/arch/greco/arc/security.c
+++ b/arch/greco/arc/security.c
@@ -11,6 +11,7 @@
 #include <uart.h>
 #include <sha256.h>
 #include <greco/greco_security.h>
+#include <hl.h>
 
 #define EFUSE_SYSCFG_ENABLE_VALIDATION_MASK	BIT(10)
 
@@ -102,3 +103,8 @@
 {
 	return true;
 }
+
+void preboot_security_init(void)
+{
+	preboot_secure_lbw();
+}
diff --git a/arch/greco/arm/security.c b/arch/greco/arm/security.c
index 80258d8..7b82727 100644
--- a/arch/greco/arm/security.c
+++ b/arch/greco/arm/security.c
@@ -9,13 +9,6 @@
 #include <psoc_arc0_mstr_if_core_lbw_regs.h>
 #include <psoc_arc1_mstr_if_core_hbw_regs.h>
 #include <psoc_arc1_mstr_if_core_lbw_regs.h>
-#include <dcore0_tpcif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
-#include <dcore1_tpcif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
-#include <dcore0_mmeif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
-#include <dcore1_mmeif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
-#include <pcie_elbi_rr_mstr_if_rr_shrd_lbw_regs.h>
-#include <dcon0_lbw_rtr_if_mstr_if_rr_prvt_lbw_regs.h>
-#include <dcon0_lbw_rtr_if_mstr_if_rr_shrd_lbw_regs.h>
 #include <dcore0_tpcif_rtr0_mstr_if_rr_prvt_hbw_regs.h>
 #include <dcore1_tpcif_rtr0_mstr_if_rr_prvt_hbw_regs.h>
 #include <dcore0_mmeif_rtr0_mstr_if_rr_prvt_hbw_regs.h>
@@ -24,55 +17,8 @@
 #include <dcon0_hbw_rtr_if0_mstr_if_rr_shrd_hbw_regs.h>
 #include <dcon0_hbw_rtr_if1_mstr_if_rr_prvt_hbw_regs.h>
 #include <dcon0_hbw_rtr_if1_mstr_if_rr_shrd_hbw_regs.h>
-
+#include "hl.h"
 #include "uart.h"
-
-struct preboot_sec_rrs_cfg {
-	u32 instances_num;
-	u32 instances_dist;
-	u32 min_short_breg;
-	u32 max_short_breg;
-	u32 min_full_breg;
-	u32 min_full_breg_hi;
-	u32 max_full_breg;
-	u32 max_full_breg_hi;
-	u32 hit_wr_breg;
-	u32 hit_rd_breg;
-};
-
-/*
- * enum preboot_sec_common_engine_type classifies the interfaces with
- * respect to the initiator engine. TPC, MME and ELBI have unique
- * interfaces, while others (DEC/ENC/PSOC/CPU etc) share the common
- * DCONx interfaces which here is prefixed as 'COMMON_IF_'.
- *
- * ----------------------------------
- *        LBW       |      HBW
- * ----------------------------------
- * TPC_MME_IF       |TPC_MME_IF
- * COMMON_IF_PRVT   |COMMON_IF_PRVT
- * COMMON_IF_SHRD   |COMMON_IF_SHRD
- * ELBI             |COMMON_IF2
- *                  |COMMON_IF_SHRD2
- * ----------------------------------
- */
-enum preboot_sec_engine_type {
-	TPC_MME_IF,
-	COMMON_IF_PRVT,
-	COMMON_IF_SHRD,
-	ELBI_OR_COMMON_IF2,
-	COMMON_IF_SHRD2,
-};
-
-enum greco_sec_if_lbw_rrs {
-	LBW_RR_PCIE_MSI,
-	LBW_MAX_RR_IDX = 18,
-};
-
-enum greco_sec_if_hbw_rrs {
-	HBW_RR_BOOTROM,
-	HBW_MAX_RR_IDX = 10,
-};
 
 #define EN_ARC_SEC_PRIV		0x3
 
@@ -88,63 +34,6 @@
 	regmap_writel(EN_ARC_SEC_PRIV,
 		 mmPSOC_ARC1_MSTR_IF_CORE_LBW_AXPROT_OVRD);
 }
-
-#define SHIFT_4K_POS		12
-#define TPC_MME_IF_NUM		16
-#define TPC_MME_IF_DIST		0x8000
-#define ELBI_IF_NUM		1
-#define ELBI_IF_DIST		0x0
-#define COMMON_PRVT_NUM		4
-#define COMMON_PRVT_DIST	0x10000
-#define COMMON_SHRD_NUM		2
-#define COMMON_SHRD_DIST	(COMMON_PRVT_DIST * 2)
-#define COMMON_HBW_PRVT_NUM	COMMON_PRVT_NUM
-#define COMMON_HBW_PRVT_DIST	COMMON_PRVT_DIST
-#define COMMON_HBW_SHRD_NUM	COMMON_SHRD_NUM
-#define COMMON_HBW_SHRD_DIST	COMMON_SHRD_DIST
-
-static struct preboot_sec_rrs_cfg preboot_sec_lbw_rrs_tbl[] = {
-	[TPC_MME_IF] = {
-		.instances_num = TPC_MME_IF_NUM,
-		.instances_dist = TPC_MME_IF_DIST,
-		.min_short_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MIN_SHORT_0,
-		.max_short_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MAX_SHORT_0,
-		.min_full_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MIN_0,
-		.max_full_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MAX_0,
-		.hit_wr_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AW,
-		.hit_rd_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AR,
-	},
-	[COMMON_IF_PRVT] = {
-		.instances_num = COMMON_PRVT_NUM,
-		.instances_dist = COMMON_PRVT_DIST,
-		.min_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MIN_SHORT_0,
-		.max_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MAX_SHORT_0,
-		.min_full_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MIN_0,
-		.max_full_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MAX_0,
-		.hit_wr_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AW,
-		.hit_rd_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AR,
-	},
-	[COMMON_IF_SHRD] = {
-		.instances_num = COMMON_SHRD_NUM,
-		.instances_dist = COMMON_SHRD_DIST,
-		.min_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
-		.max_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
-		.min_full_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_0,
-		.max_full_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_0,
-		.hit_wr_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AW,
-		.hit_rd_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
-	},
-	[ELBI_OR_COMMON_IF2] = {
-		.instances_num = ELBI_IF_NUM,
-		.instances_dist = ELBI_IF_DIST,
-		.min_short_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
-		.max_short_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
-		.min_full_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_0,
-		.max_full_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_0,
-		.hit_wr_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AW,
-		.hit_rd_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
-	},
-};
 
 static struct preboot_sec_rrs_cfg preboot_sec_hbw_rrs_tbl[] = {
 	[TPC_MME_IF] = {
@@ -239,58 +128,6 @@
 	regmap_writel(hit | BIT(rr_idx), rr_dest);
 }
 
-#define MAX_LBW_SHORT_RR 14
-
-static void preboot_secure_lbw_rr_gen(int rr_idx, u32 min, u32 max,
-		enum preboot_sec_engine_type type, bool is_short)
-{
-	int i;
-	u32 min_breg, max_breg, rr_calc;
-	struct preboot_sec_rrs_cfg *cfg = &preboot_sec_lbw_rrs_tbl[type];
-
-	if (is_short) {
-		rr_calc = rr_idx;
-		min_breg = cfg->min_short_breg;
-		max_breg = cfg->max_short_breg;
-	} else {
-		rr_calc = rr_idx - MAX_LBW_SHORT_RR;
-		min_breg = cfg->min_full_breg;
-		max_breg = cfg->max_full_breg;
-	}
-
-	for (i = 0; i < cfg->instances_num; i++) {
-		preboot_secure_update_rr_reg(rr_calc, cfg->instances_dist * i,
-				min_breg, max_breg, min, max);
-
-		preboot_secure_update_hit_reg(rr_idx, cfg, i, false);
-	}
-}
-
-static void preboot_secure_lbw_rr_wrap(int rr_idx, u32 min, u32 max, bool is_short)
-{
-	int i;
-
-	for (i = TPC_MME_IF; i <= ELBI_OR_COMMON_IF2; i++)
-		preboot_secure_lbw_rr_gen(rr_idx, min, max, (enum preboot_sec_engine_type)i,
-				is_short);
-}
-
-__unused static void preboot_secure_lbw_rr_full(int rr_idx, u64 min64, u64 max64)
-{
-	u32 min = min64 & ~HL_CFG_BASE;
-	u32 max = max64 & ~HL_CFG_BASE;
-
-	preboot_secure_lbw_rr_wrap(rr_idx, min, max, false);
-}
-
-static void preboot_secure_lbw_rr_short(int rr_idx, u64 min64, u64 max64)
-{
-	u32 min = (min64 & ~HL_CFG_BASE) >> SHIFT_4K_POS;
-	u32 max = (max64 & ~HL_CFG_BASE) >> SHIFT_4K_POS;
-
-	preboot_secure_lbw_rr_wrap(rr_idx, min, max, true);
-}
-
 #define MAX_HBW_SHORT_RR	6
 #define MAX_HBW_READ_RRS	7
 #define BITS_IN_REG		32
@@ -360,33 +197,6 @@
 
 	for (i = TPC_MME_IF; i < COMMON_IF_SHRD2; i++)
 		preboot_secure_hbw_rr_short_if(rr, min, max, (enum preboot_sec_engine_type)i);
-}
-
-static void preboot_secure_lbw(void)
-{
-	u64 min, max;
-	/*
-	 * There are 18(0..17 rr idx) RRs for LBW.
-	 * The classification and apis are as listed below:
-	 *
-	 * 1) 14(0..13 rr idx) Regions with 4KB granularity
-	 * API: preboot_secure_lbw_rr_short(rr idx, min addr, max addr)
-	 *
-	 * 2) 4(14..17 rr idx) Regions with full address range.
-	 * API: preboot_secure_lbw_rr_full(rr idx, min addr, max addr)
-	 *
-	 * rr idx type : greco_sec_if_lbw_rrs
-	 * NOTE: The apis listed above will block all read and write
-	 * transactions for the input address range.
-	 */
-
-	/*
-	 * Workaround for H5 Errata: GOYA2_0100
-	 * Block PCIE_MSI : RR(LBW_RR_PCIE_MSI)
-	 */
-	min = mmPCIE_MSI_BASE;
-	max = mmPCIE_MSI_BASE + PCIE_MSI_SECTION;
-	preboot_secure_lbw_rr_short(LBW_RR_PCIE_MSI, min, max);
 }
 
 static void preboot_secure_hbw(void)
diff --git a/arch/greco/common/common.c b/arch/greco/common/common.c
index d9c705d..54360a4 100644
--- a/arch/greco/common/common.c
+++ b/arch/greco/common/common.c
@@ -50,6 +50,13 @@
 #include <pcie_aux_masks.h>
 #include <gic_regs.h>
 #include <greco_irq.h>
+#include <dcore0_tpcif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
+#include <dcore1_tpcif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
+#include <dcore0_mmeif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
+#include <dcore1_mmeif_rtr0_mstr_if_rr_prvt_lbw_regs.h>
+#include <pcie_elbi_rr_mstr_if_rr_shrd_lbw_regs.h>
+#include <dcon0_lbw_rtr_if_mstr_if_rr_prvt_lbw_regs.h>
+#include <dcon0_lbw_rtr_if_mstr_if_rr_shrd_lbw_regs.h>
 
 #define mmPCIE_WRAP_VUART_TX_DATA       0xC15F60
 #define mmPCIE_WRAP_VUART_TX_STATUS     0xC15F64
@@ -57,6 +64,41 @@
 #define UNIT_RST_N_L_MSK		0xF0000
 #define UNIT_RST_N_M_MSK		0x0
 #define UNIT_RST_N_H_MSK		0x0
+
+static struct preboot_sec_rrs_cfg preboot_sec_lbw_rrs_tbl[] = {
+	[TPC_MME_IF] = {
+		.instances_num = TPC_MME_IF_NUM,
+		.instances_dist = TPC_MME_IF_DIST,
+		.min_short_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MAX_SHORT_0,
+		.hit_wr_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCORE0_TPCIF_RTR0_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AR,
+	},
+	[COMMON_IF_PRVT] = {
+		.instances_num = COMMON_PRVT_NUM,
+		.instances_dist = COMMON_PRVT_DIST,
+		.min_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_RANGE_MAX_SHORT_0,
+		.hit_wr_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_PRVT_LBW_PRIV_HIT_AR,
+	},
+	[COMMON_IF_SHRD] = {
+		.instances_num = COMMON_SHRD_NUM,
+		.instances_dist = COMMON_SHRD_DIST,
+		.min_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
+		.hit_wr_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmDCON0_LBW_RTR_IF_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
+	},
+	[ELBI_OR_COMMON_IF2] = {
+		.instances_num = ELBI_IF_NUM,
+		.instances_dist = ELBI_IF_DIST,
+		.min_short_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MIN_SHORT_0,
+		.max_short_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_RANGE_MAX_SHORT_0,
+		.hit_wr_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AW,
+		.hit_rd_breg = mmPCIE_ELBI_RR_MSTR_IF_RR_SHRD_LBW_PRIV_HIT_AR,
+	},
+};
 
 void hl_ts_reset(void)
 {
@@ -850,3 +892,84 @@
 	regmap_writel(LBW_DRAIN_TIMEOUT, mmPCIE_WRAP_LBW_DRAIN_TIMEOUT);
 	regmap_writel(1, mmPCIE_WRAP_LBW_DRAIN_CFG);
 }
+
+static void preboot_secure_update_rr_reg(int rr_idx, u32 inst_offset, u32 min_reg, u32 max_reg,
+		u32 min, u32 max)
+{
+	u32 rr_base, rr_dest;
+
+	rr_base = rr_idx * sizeof(u32) + inst_offset;
+	rr_dest = rr_base + min_reg;
+	regmap_writel(min, rr_dest);
+	rr_dest = rr_base + max_reg;
+	regmap_writel(max, rr_dest);
+}
+
+static void preboot_secure_update_hit_reg(int rr_idx, struct preboot_sec_rrs_cfg *cfg, int inst_num,
+		bool write_only)
+{
+	u32 rr_base, rr_dest, hit;
+	/* Enable (hit) RR */
+	rr_base = cfg->instances_dist * inst_num;
+	rr_dest = rr_base + cfg->hit_wr_breg;
+	hit = regmap_readl(rr_dest);
+	regmap_writel(hit | BIT(rr_idx), rr_dest);
+
+	if (write_only)
+		return;
+
+	rr_dest = rr_base + cfg->hit_rd_breg;
+	hit = regmap_readl(rr_dest);
+	regmap_writel(hit | BIT(rr_idx), rr_dest);
+}
+
+static void preboot_secure_lbw_rr_gen(int rr_idx, u32 min, u32 max,
+		enum preboot_sec_engine_type type)
+{
+	int i;
+	struct preboot_sec_rrs_cfg *cfg = &preboot_sec_lbw_rrs_tbl[type];
+
+	for (i = 0; i < cfg->instances_num; i++) {
+		preboot_secure_update_rr_reg(rr_idx, cfg->instances_dist * i,
+				cfg->min_short_breg, cfg->max_short_breg, min, max);
+
+		preboot_secure_update_hit_reg(rr_idx, cfg, i, false);
+	}
+}
+
+static void preboot_secure_lbw_rr_short(int rr_idx, u64 min64, u64 max64)
+{
+	u32 min = (min64 & ~HL_CFG_BASE) >> SHIFT_4K_POS;
+	u32 max = (max64 & ~HL_CFG_BASE) >> SHIFT_4K_POS;
+	int i;
+
+	for (i = TPC_MME_IF; i <= ELBI_OR_COMMON_IF2; i++)
+		preboot_secure_lbw_rr_gen(rr_idx, min, max, (enum preboot_sec_engine_type)i);
+}
+
+void preboot_secure_lbw(void)
+{
+	u64 min, max;
+	/*
+	 * There are 18(0..17 rr idx) RRs for LBW.
+	 * The classification and apis are as listed below:
+	 *
+	 * 1) 14(0..13 rr idx) Regions with 4KB granularity
+	 * API: preboot_secure_lbw_rr_short(rr idx, min addr, max addr)
+	 *
+	 * 2) 4(14..17 rr idx) Regions with full address range.
+	 * API: preboot_secure_lbw_rr_full(rr idx, min addr, max addr)
+	 *
+	 * rr idx type : greco_sec_if_lbw_rrs
+	 * NOTE: The apis listed above will block all read and write
+	 * transactions for the input address range.
+	 */
+
+	/*
+	 * Workaround for H5 Errata: GOYA2_0100
+	 * Block PCIE_MSI : RR(LBW_RR_PCIE_MSI)
+	 */
+	min = mmPCIE_MSI_BASE;
+	max = mmPCIE_MSI_BASE + PCIE_MSI_SECTION;
+	preboot_secure_lbw_rr_short(LBW_RR_PCIE_MSI, min, max);
+}
diff --git a/arch/greco/common/hl.h b/arch/greco/common/hl.h
index 53982e6..ed0458d 100644
--- a/arch/greco/common/hl.h
+++ b/arch/greco/common/hl.h
@@ -40,6 +40,10 @@
 #define ARC0_MEM_SIZE			(ARC0_ICCM_SIZE + ARC0_DCCM_SIZE)
 #define ARC1_MEM_SIZE			(ARC1_ICCM_SIZE + ARC1_DCCM_SIZE)
 
+#define BOOT_ROM_BASE			0x7FFB040000ull
+#define BOOT_ROM_SIZE			SZ_64K
+#define FW_SRAM_SIZE			SZ_64K
+
 enum hl_rst_src {
 	HL_COLD_RST,
 	HL_MANUAL_RST,
@@ -51,9 +55,71 @@
 	HL_FLR_RST = 9,
 };
 
+#define SHIFT_4K_POS		12
+#define TPC_MME_IF_NUM		16
+#define TPC_MME_IF_DIST		0x8000
+#define ELBI_IF_NUM		1
+#define ELBI_IF_DIST		0x0
+#define COMMON_PRVT_NUM		4
+#define COMMON_PRVT_DIST	0x10000
+#define COMMON_SHRD_NUM		2
+#define COMMON_SHRD_DIST	(COMMON_PRVT_DIST * 2)
+#define COMMON_HBW_PRVT_NUM	COMMON_PRVT_NUM
+#define COMMON_HBW_PRVT_DIST	COMMON_PRVT_DIST
+#define COMMON_HBW_SHRD_NUM	COMMON_SHRD_NUM
+#define COMMON_HBW_SHRD_DIST	COMMON_SHRD_DIST
+
+struct preboot_sec_rrs_cfg {
+	u32 instances_num;
+	u32 instances_dist;
+	u32 min_short_breg;
+	u32 max_short_breg;
+	u32 min_full_breg;
+	u32 min_full_breg_hi;
+	u32 max_full_breg;
+	u32 max_full_breg_hi;
+	u32 hit_wr_breg;
+	u32 hit_rd_breg;
+};
+
+/*
+ * enum preboot_sec_common_engine_type classifies the interfaces with
+ * respect to the initiator engine. TPC, MME and ELBI have unique
+ * interfaces, while others (DEC/ENC/PSOC/CPU etc) share the common
+ * DCONx interfaces which here is prefixed as 'COMMON_IF_'.
+ *
+ * ----------------------------------
+ *        LBW       |      HBW
+ * ----------------------------------
+ * TPC_MME_IF       |TPC_MME_IF
+ * COMMON_IF_PRVT   |COMMON_IF_PRVT
+ * COMMON_IF_SHRD   |COMMON_IF_SHRD
+ * ELBI             |COMMON_IF2
+ *                  |COMMON_IF_SHRD2
+ * ----------------------------------
+ */
+enum preboot_sec_engine_type {
+	TPC_MME_IF,
+	COMMON_IF_PRVT,
+	COMMON_IF_SHRD,
+	ELBI_OR_COMMON_IF2,
+	COMMON_IF_SHRD2,
+};
+
+enum greco_sec_if_lbw_rrs {
+	LBW_RR_PCIE_MSI,
+	LBW_MAX_RR_IDX = 18,
+};
+
+enum greco_sec_if_hbw_rrs {
+	HBW_RR_BOOTROM,
+	HBW_MAX_RR_IDX = 10,
+};
+
 void run_arc(u8 id);
 void halt_n_reset_arc(u8 id);
 void axi_enable_drain(void);
 void init_arc_mem(u8 id, u64 addr, u32 size);
+void preboot_secure_lbw(void);
 
 #endif /* ARCH_GRECO_COMMON_HL_H_ */
